### ISSUE TITLE
Combining scope-level and record-level

### ISSUE THREAD

**Issue by danielbecroft** [thepirat000/Audit.NET#339]
Apologies if this is not the best forum, but I have a question as to the usage of Audit.NET.

I'd like to be able to audit entries at an "event" level, to track the start/end times, and changes associated with, a given batch operation (ie "BulkUpdate", "ImportFile", etc), but also track the individual record/model level changes as well.

So given a batch event, see all the model changes associated with that:
```
BulkUpdate Id:1000, 27/07/2020 
- Order Id:1, update status to "Complete"
- Order Id:2, update status to "Payment Rejected"
- Order Id:3, update status to "Complete
```

but then also track all the changes for a given item and surface these via APIs as well:

```
Order Id:1
- 27/07/2020 - Status set to "Created"
- 27/07/2020 - Status set to "Pending"
- 27/07/2020 - Status set to "Complete" (BulkUpdate Id:1000)
```

Is something like this possible with Audit.NET? Would it require tracking data changes separately?

**Comment by thepirat000** [comment]
depends on how are you doing the operations 
it's possible if the insert/update operations are controlled via .NET code

**Comment by danielbecroft** [comment]
Okay thanks @thepirat000 . Is there a high-level idea of how that might be attempted?

**Comment by danielbecroft** [comment]
We've attempted to implement something for this, but we have some .... questions/concerns around the implementation. I'm not sure if these are known, intended, or we just missed the documentation.

The first thing we noticed, is that, if we use the `AuditScope` for the bulk operation, only data changes associated with the `Target` are tracked against this event. For example:
```
using var scope = AuditScope.Create(new AuditScopeOptions() { EventType = "Order:BulkUpdate" });

foreach (var order in _context.Orders)
{
    if (_service.UpdateRequired(order))
        _service.UpdateOrder(order);
}
```

Using this approach, the changes to the `order` instance are not tracked. We would need to know the collection of entries that will be updated, and ensure they get set as the `Target` property. But this isn't doable.

So our current approach that we're going to try is:
1. Use the `AuditDbContext` rather than `DbContext`
2. Create an `AuditScope` for the bulk-type operations, or any operation we need to apply a business-level operation to. As a part of this, we generate a `CorrelationId` (unique)
3. In the `DbContext`, override the `OnScopeCreated`, and set the `CorrelationId` custom field using the value generated at (2).

This allows us to:
1. Find all the top-level events (base on the `EventType` column), and then all the underlying changes that each of those produced (using a `jsonb` query on `CorrelationId` in the `Data` colum); and
2. Find all the events for a record by filtering directly on the `Data` column using the Entity framework model

Is this the correct approach? Are there anything that we would need to take care with?
